crates/ruff/src/rules/ruff/typing.rs

```diff
@@ -0,0 +1,337 @@
+use rustpython_parser::ast::{self, Constant, Expr, Operator};
+
+use ruff_python_ast::call_path::CallPath;
+use ruff_python_ast::source_code::Locator;
+use ruff_python_ast::typing::parse_type_annotation;
+use ruff_python_semantic::SemanticModel;
+use ruff_python_stdlib::sys::is_known_standard_library;
+
+/// Custom iterator to collect all the `|` separated expressions in a PEP 604
+/// union type.
+struct PEP604UnionIterator<'a> {
+    stack: Vec<&'a Expr>,
+}
+
+impl<'a> PEP604UnionIterator<'a> {
+    fn new(expr: &'a Expr) -> Self {
+        Self { stack: vec![expr] }
+    }
+}
+
+impl<'a> Iterator for PEP604UnionIterator<'a> {
+    type Item = &'a Expr;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(expr) = self.stack.pop() {
+            match expr {
+                Expr::BinOp(ast::ExprBinOp {
+                    left,
+                    op: Operator::BitOr,
+                    right,
+                    ..
+                }) => {
+                    self.stack.push(left);
+                    self.stack.push(right);
+                }
+                _ => return Some(expr),
+            }
+        }
+        None
+    }
+}
+
+/// Returns `true` if the given call path is a known type.
+///
+/// A known type is either a builtin type, any object from the standard library,
+/// or a type from the `typing_extensions` module.
+fn is_known_type(call_path: &CallPath, minor_version: u32) -> bool {
+    match call_path.as_slice() {
+        ["" | "typing_extensions", ..] => true,
+        [module, ..] => is_known_standard_library(minor_version, module),
+        _ => false,
+    }
+}
+
+#[derive(Debug)]
+enum TypingTarget<'a> {
+    None,
+    Any,
+    Object,
+    ForwardReference(Expr),
+    Union(Vec<&'a Expr>),
+    Literal(Vec<&'a Expr>),
+    Optional(&'a Expr),
+    Annotated(&'a Expr),
+
+    /// Special type used to represent an unknown type (and not a typing target).
+    /// This could be a type alias. It's also used when parsing a forward
+    /// reference fails.
+    ///
+    /// Note that this is not the same as returning `None` in `try_from_expr` as
+    /// `None` means that it's a known type but we don't know which one.
+    Unknown,
```

Comment:
Here's the difference between `None` and `Some(TypingTarget::Unknown)`:

| `None` | `Some(TypingTarget::Unknown)` |
|--------|--------|
| Known type but not a `TypingTarget` | Unknown type and not a `TypingTarget` |
| When `ExprSubscript.slice` is not a `ExprTuple` | Parsing fails for forward references |
| **Flag violation** | **Do NOT flag violation** |

My thought is to just remove the `None` variant. This means in failure we would _not_ flag violation (currently we would). Any thoughts? \cc @charliermarsh

Reply:
Could we replace `None` with `Known`, and use `Result` to return an error when "Parsing fails for forward references" and "When `ExprSubscript.slice` is not a `ExprTuple`"? Even if we can't use `Result`, I think it makes sense to encode "Known type but not a `TypingTarget`" as its own variant. It's confusing that it's represented by `None`.

Reply:
> Even if we can't use `Result`, I think it makes sense to encode "Known type but not a `TypingTarget`" as its own variant. It's confusing that it's represented by `None`.

This makes sense. Thanks!

I'm going with `Option<...>` and not flagging any violation in case of `None`.